### ISSUE TITLE
functional-tester: compact request times out

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#5440]
```
2016-05-24 00:29:28.720576 I | etcd-tester: [round#77] compacting storage at 110522767 (current revision 110532767)
2016-05-24 00:29:28.720582 I | etcd-tester: [DEBUG] compactKV grpc.Dial starting (rev 110522767, endpoint 10.240.0.17:2379)
2016-05-24 00:29:28.720615 I | etcd-tester: [DEBUG] compactKV kvc.Compact starting (rev 110522767, endpoint 10.240.0.17:2379)
2016-05-24 00:29:58.721900 I | etcd-tester: [DEBUG] compactKV kvc.Compact error rpc error: code = 4 desc = context deadline exceeded (rev 110522767, endpoint 10.240.0.17:2379, took 30.001265428s)
2016-05-24 00:29:58.721955 I | etcd-tester: [DEBUG] compactKV grpc.Dial starting (rev 110522767, endpoint 10.240.0.18:2379)
2016-05-24 00:29:58.721998 I | etcd-tester: [DEBUG] compactKV kvc.Compact starting (rev 110522767, endpoint 10.240.0.18:2379)
2016-05-24 00:29:58.736854 I | etcd-tester: 10.240.0.18:2379 is already compacted with 110522767 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-24 00:29:58.736898 I | etcd-tester: [DEBUG] compactKV grpc.Dial starting (rev 110522767, endpoint 10.240.0.19:2379)
2016-05-24 00:29:58.736945 I | etcd-tester: [DEBUG] compactKV kvc.Compact starting (rev 110522767, endpoint 10.240.0.19:2379)
2016-05-24 00:29:58.754581 I | etcd-tester: 10.240.0.19:2379 is already compacted with 110522767 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-24 00:29:58.754615 I | etcd-tester: [round#77] compactKV error (rpc error: code = 4 desc = context deadline exceeded)
```

The error happens from the node 1 (`10.240.0.17:2379`):

```
2016-05-24 00:29:28.720615 I | etcd-tester: [DEBUG] compactKV kvc.Compact starting (rev 110522767, endpoint 10.240.0.17:2379)
2016-05-24 00:29:58.721900 I | etcd-tester: [DEBUG] compactKV kvc.Compact error rpc error: code = 4 desc = context deadline exceeded (rev 110522767, endpoint 10.240.0.17:2379, took 30.001265428s)
```

1. The compact request has timeout of 30 seconds (https://github.com/coreos/etcd/blob/master/tools/functional-tester/etcd-tester/cluster.go#L335).
2. Since the request is sent with `Physical: true`, the server does not return until the compaction is actually done (https://github.com/coreos/etcd/blob/master/mvcc/kvstore.go#L294).
3. And the node 1 log shows that it took almost 30 seconds to finish the compaction, which leads to context deadline timeout.

```
2016-05-24 00:29:28.729809 I | mvcc: store.index: compact 110522767
2016-05-24 00:29:58.715819 I | mvcc: finished scheduled compaction at 110522767 (took 29.784187883s)
```

To summarize, the compact request went through, but the compaction took as much as timeout (30-sec).

I will try longer timeout and see if we still get this error.

**Comment by gyuho** [comment]
```
Compacting 1565040 entries took 22.224533451s (71138 per second)
Compacting 1370733 entries took 20.072420034s (68536 per second)
```

Been running with longer timeout without any failure over 200 rounds. So pretty sure that configuring the right compaction timeout would resolve this. We need to cache the previous index and get the number of entries to compact, and apply different timeouts case by case.

For our testing environments, compaction rate is about 65,000 entries per second. I have a patch but need to resolve other issues in functional tester. Will send a PR that addresses it all.

**Comment by gyuho** [comment]
```
2016-05-27 00:49:06.768607 I | etcd-tester: [round#72] compacting 1851204 entries (timeout 47s)
2016-05-27 00:49:06.768614 I | etcd-tester: [round#72] compacting storage (current revision 92395961, compact revision 92385961)
2016-05-27 00:49:34.028082 I | etcd-tester: 10.240.0.18:2379 is already compacted with 92385961 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-27 00:49:39.029746 I | etcd-tester: [round#72] compactKV error (rpc error: code = 13 desc = context deadline exceeded)
2016-05-27 00:49:39.029803 I | etcd-tester: [round#72] cleaning up...

# 10.240.0.17:2379
2016-05-27 00:49:06.780094 I | mvcc: store.index: compact 92385961
2016-05-27 00:49:33.080612 I | mvcc: finished scheduled compaction at 92385961 (took 26.136602748s)

# 10.240.0.17:2379
2016-05-27 00:49:06.778525 I | mvcc: store.index: compact 92385961
2016-05-27 00:49:34.019549 I | mvcc: finished scheduled compaction at 92385961 (took 26.996168284s)

# 10.240.0.20:2379
2016-05-27 00:49:06.778508 I | mvcc: store.index: compact 92385961
2016-05-27 00:49:32.959823 I | mvcc: finished scheduled compaction at 92385961 (took 26.006928121s)
```

Happen again. TImeout seemed to happen for the first endpoint, but we don't log endpoints, so it's hard to tell which was timing out.

**Comment by gyuho** [comment]
Closing. Will revisit when it happens again.

**Comment by gyuho** [comment]
Got this again...

```
2016-06-03 14:19:21.395241 I | etcd-tester: [round#136] compacting 1012186 entries (timeout 30s)
2016-06-03 14:19:21.395253 I | etcd-tester: [round#136] compacting storage (current revision 113543120, compact revision 113533120)
2016-06-03 14:19:21.395286 I | etcd-tester: [compact kv #0] starting (endpoint 10.240.0.2:2379)
2016-06-03 14:19:37.991977 I | etcd-tester: [compact kv #0] done (endpoint 10.240.0.2:2379)
2016-06-03 14:19:37.992071 I | etcd-tester: [compact kv #1] starting (endpoint 10.240.0.3:2379)
2016-06-03 14:19:42.993379 I | etcd-tester: [compact kv #1] error rpc error: code = 13 desc = context deadline exceeded (endpoint 10.240.0.3:2379)
2016-06-03 14:19:42.993465 I | etcd-tester: [compact kv #2] starting (endpoint 10.240.0.4:2379)
2016-06-03 14:19:43.013586 I | etcd-tester: [compact kv #2] already compacted (endpoint 10.240.0.4:2379)
2016-06-03 14:19:43.013652 I | etcd-tester: [compact kv #2] done (endpoint 10.240.0.4:2379)
2016-06-03 14:19:43.013692 I | etcd-tester: [round#136] cleaning up...
2016-06-03 14:19:48.942654 W | etcd-tester: [round#136] functional-tester returning with error (rpc error: code = 13 desc = context deadline exceeded)
```

Compact request was successfully replicated in all 3 nodes, but agent2 was taking longer (I assume, because there is no `mvcc: compaction done` message in the log). And another weird thing is the context timeout was 30 seconds, but the compact request returns in only 5 seconds...

I will add more logs in `mvcc` package, to keep track of the progress of compaction, and redeploy the functional cluster. The logs from master branch doesn't help a lot to debug this.

**Comment by gyuho** [comment]
```
# tester
#
2016-06-08 12:38:21.654037 I | etcd-tester: [round#117] compacting 1456983 entries (timeout 39s)
2016-06-08 12:38:21.654078 I | etcd-tester: [round#117] compacting storage (current revision 7437246, compact revision 7427246)
2016-06-08 12:38:21.654114 I | etcd-tester: [compact kv #0] starting (endpoint 10.240.0.13:2379)
2016-06-08 12:39:00.654284 I | etcd-tester: [compact kv #0] error rpc error: code = 4 desc = context deadline exceeded (endpoint 10.240.0.13:2379)
2016-06-08 12:39:00.654383 I | etcd-tester: [compact kv #1] starting (endpoint 10.240.0.14:2379)
2016-06-08 12:39:00.668267 I | etcd-tester: [compact kv #1] already compacted (endpoint 10.240.0.14:2379)
2016-06-08 12:39:00.668308 I | etcd-tester: [compact kv #1] done (endpoint 10.240.0.14:2379)
2016-06-08 12:39:00.668350 I | etcd-tester: [compact kv #2] starting (endpoint 10.240.0.15:2379)
2016-06-08 12:39:00.691179 I | etcd-tester: [compact kv #2] already compacted (endpoint 10.240.0.15:2379)
2016-06-08 12:39:00.691241 I | etcd-tester: [compact kv #2] done (endpoint 10.240.0.15:2379)
2016-06-08 12:39:00.691276 I | etcd-tester: [round#117] cleaning up...

# 10.240.0.13:2379
#
# compaction starts (39s to finish)
2016-06-08 12:38:21.709978 I | mvcc: store.index: compact 7427246
2016-06-08 12:38:22.183461 I | etcdserver: start to snapshot (applied: 7453467, lastsnap: 7443465)
2016-06-08 12:38:22.192653 I | etcdserver: saved snapshot at index 7453467
2016-06-08 12:38:22.192825 I | etcdserver: compacted raft log at 7448467
...
2016-06-08 12:38:26.541215 I | etcdserver: start to snapshot (applied: 7533945, lastsnap: 7523888)
2016-06-08 12:38:26.548652 I | etcdserver: saved snapshot at index 7533945
2016-06-08 12:38:26.548815 I | etcdserver: compacted raft log at 7528945
# snapshot
2016-06-08 12:38:27.687988 I | rafthttp: receiving database snapshot [index:7550738, from d323b71a635a0445] ...
2016-06-08 12:38:29.505866 I | fileutil: purged file agent.etcd/member/snap/000000000000002e-00000000006bf836.snap successfully
...
2016-06-08 12:38:29.507407 I | fileutil: purged file agent.etcd/member/snap/000000000000002e-00000000007230ca.snap successfully
2016-06-08 12:38:29.523115 I | fileutil: purged file agent.etcd/member/wal/000000000000000c-00000000004bcc48.wal successfully
2016-06-08 12:38:29.612736 I | snap: saved database snapshot to disk [total bytes: 343760896]
2016-06-08 12:38:29.612814 I | rafthttp: received and saved database snapshot [index: 7550738, from: d323b71a635a0445] successfully
2016-06-08 12:38:29.613000 I | raft: 437b8abd875de8e5 [commit: 7536016, lastindex: 7536016, lastterm: 46] starts to restore snapshot [index: 7550738, term: 46]
2016-06-08 12:38:29.613035 I | raft: log [committed=7536016, applied=7536016, unstable.offset=7536017, len(unstable.Entries)=0] starts to restore snapshot [index: 7550738, term: 46]
2016-06-08 12:38:29.613055 I | raft: 437b8abd875de8e5 restored progress of 437b8abd875de8e5 [next = 7550739, match = 7550738, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-08 12:38:29.613065 I | raft: 437b8abd875de8e5 restored progress of d323b71a635a0445 [next = 7550739, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-08 12:38:29.613076 I | raft: 437b8abd875de8e5 restored progress of f33b1e0cca75eac9 [next = 7550739, match = 0, state = ProgressStateProbe, waiting = false, pendingSnapshot = 0]
2016-06-08 12:38:29.613084 I | raft: 437b8abd875de8e5 [commit: 7550738] restored snapshot [index: 7550738, term: 46]
2016-06-08 12:38:29.613222 I | etcdserver: applying snapshot at index 7533945...
2016-06-08 12:38:29.617338 I | etcdserver: raft applied incoming snapshot at index 7550738
# restoring from snapshot starts
2016-06-08 12:38:29.639266 I | etcdserver: restoring mvcc store...
2016-06-08 12:38:29.658983 I | wal: segmented wal file agent.etcd/member/wal/0000000000000012-00000000007371b3.wal is created
2016-06-08 12:38:53.696581 I | wal: segmented wal file agent.etcd/member/wal/0000000000000013-0000000000798286.wal is created
2016-06-08 12:38:59.507666 I | fileutil: purged file agent.etcd/member/snap/000000000000002e-00000000007257e9.snap successfully
2016-06-08 12:38:59.533436 I | fileutil: purged file agent.etcd/member/wal/000000000000000d-000000000051dd0c.wal successfully
2016-06-08 12:38:59.542172 I | fileutil: purged file agent.etcd/member/wal/000000000000000e-000000000058fa0a.wal successfully
2016-06-08 12:39:00.694335 N | osutil: received terminated signal, shutting down...


# To summarize,
# compaction should have done by 12:39:00
# but stresser triggers snapshot at 12:38:29
# and the restoring from snapshot didn't finish by 12:39:00
```

The root cause was https://github.com/coreos/etcd/issues/5317.
